client/state/data-layer/wpcom-http/utils.js

```diff
@@ -161,3 +166,77 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) =
 
 	return initiator( store, action );
 };
+
+/*
+ * Converts an application-level Calypso action that's handled by the data-layer middleware
+ * into a low-level action. For example, HTTP request that's being initiated, or a response
+ * action with a `meta.dataLayer` property.
+ */
+const createRequestAction = ( options, action ) => {
+	const {
+		initiate = noop,
+		onSuccess = noop,
+		onError = noop,
+		onProgress = noop,
+		fromApi = identity,
+	} = options;
+
+	const data = getData( action );
+	if ( data ) {
+		try {
+			return onSuccess( action, fromApi( data ) );
+		} catch ( err ) {
+			return onError( action, err );
+		}
+	}
+
+	const error = getError( action );
+	if ( error ) {
+		return onError( action, error );
+	}
+
+	const progress = getProgress( action );
+	if ( progress ) {
+		return onProgress( action, progress );
+	}
+
+	const initiator = initiate( action );
+	if ( ! initiator ) {
+		warn(
+			"The `initiate` handler didn't return any action to dispatch: you probably wanted to initiate a request?"
+		);
+	}
+
+	return Object.assign( {}, initiator, {
+		onSuccess: action,
+		onFailure: action,
+		onProgress: action,
+	} );
+};
+
+/*
+ * Create an effect handler for a particular data-layer action type.
+ */
+export const handleDataRequest = options => {
+	if ( ! options.initiate ) {
+		warn( 'initiate handler is not defined: you really should supply one' );
+	}
+
+	if ( ! options.onSuccess ) {
+		warn( 'onSuccess handler is not defined: you really should supply one' );
+	}
+
+	if ( ! options.onError ) {
+		warn( 'onError handler is not defined: please consider handling API errors correctly' );
+	}
+
+	const handler = ( store, action ) => {
+		// create the low-level action we want to dispatch
+		const requestAction = createRequestAction( options, action );
+		// dispatch the low level action (if any was created) and return the result
+		return requestAction ? store.dispatch( requestAction ) : undefined;
+	};
+
+	// Return an one-element array of handlers
+	return [ handler ];
```

Comment:
noted in my comment in the main thread, I think there is something in this we might be able to apply more generally to the handler API.

going back to the question: "is this different enough to warrant a new fundamental piece?" - could we instead update the data layer API itself with the ideas behind these changes without losing anything we want from the data layer?